core/templates/domain/feature_gating/PlatformParameterFilterObjectFactory.ts

```diff
@@ -0,0 +1,174 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Factory for creating PlatformParameterFilter domain objects.
+ */
+
+import cloneDeep from 'lodash/cloneDeep';
+
+import { Injectable } from '@angular/core';
+import { downgradeInjectable } from '@angular/upgrade/static';
+
+const constants = require('constants.ts');
+
+
+export enum PlatformParameterFilterType {
+  ServerMode = 'server_mode',
+  UserLocale = 'user_locale',
+  ClientType = 'client_type',
+  BrowserType = 'browser_type',
+  AppVersion = 'app_version',
+}
+
+export enum ServerMode {
+  Dev = 'dev',
+  Test = 'test',
+  Prod = 'prod',
+}
+
+export const ALLOWED_SERVER_MODE = [
+  ServerMode.Dev, ServerMode.Test, ServerMode.Prod].map(e => e.toString());
+
+export const SUPPORTED_SITE_LANGUAGE_IDS = (
+  constants.SUPPORTED_SITE_LANGUAGES.map((lang: {id: string}) => lang.id)
+);
+
+export const ALLOWED_CLIENT_TYPE = ['Android', 'Web'];
+
+export const ALLOWED_BROWSER_TYPE = [
+  'Chrome', 'Edge', 'Safari', 'Firefox', 'Others'];
+
+export const APP_VERSION_REGEXP = /^\d+(?:\.\d+)*$/;
+
+export const ALLOWED_OP_FOR_TYPE = {
+  [PlatformParameterFilterType.ServerMode]: ['='],
+  [PlatformParameterFilterType.UserLocale]: ['='],
+  [PlatformParameterFilterType.ClientType]: ['='],
+  [PlatformParameterFilterType.BrowserType]: ['='],
+  [PlatformParameterFilterType.AppVersion]: ['=', '<', '>', '<=', '>='],
+};
+
+export interface PlatformParameterFilterBackendDict {
+    'type': PlatformParameterFilterType;
+    'conditions': Array<[string, string]>;
+}
+
+export class PlatformParameterFilter {
+  type: PlatformParameterFilterType;
+  conditions: Array<[string, string]>;
+
+  constructor(
+      type: PlatformParameterFilterType, conditions: Array<[string, string]>) {
+    this.type = type;
+    this.conditions = conditions;
+  }
+
+  /**
+   * Creates a dict representation of the instance.
+   *
+   * @returns {PlatformParameterFilterBackendDict} - The dict representation
+   * of the instance.
+   */
+  toBackendDict(): PlatformParameterFilterBackendDict {
+    return {
+      type: this.type,
+      conditions: cloneDeep(this.conditions)
+    };
+  }
+
+  /**
+   * Validates the PlatformParameterFilter instance.
+   *
+   * @returns {string[]} The issue messages found during validation, if any.
+   */
+  validate(): string[] {
```

Comment:
I'm actually wondering--what's the intended use of this function? If it's for admin changes to the parameters, I'd expect that we can just rely on the server's validation logic. My concern is that the validation logic needs to be kept in sync between here & the server, and there's no easy way to ensure that happens.

Further, we can avoid a lot of the validation issues by ensuring the admin forms & drop-downs are fixed to all available options for given filter options, or perform simple verification (like regex checks for versions).

Reply:
Ah it makes sense..
Deleted validation methods, they will be replaced by forms & dropdowns in the admin panel.